src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+        occurrences.write_and_clear(writer)
+    }
+
+    fn collapse_multi_threaded<R>(
+        &mut self,
+        mut reader: R,
+        occurrences: &mut Occurrences,
+    ) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let nstacks_per_job = self.nstacks_per_job();
+        let nthreads = self.nthreads();
+
+        assert_ne!(nstacks_per_job, 0);
+        assert!(nthreads > 1);
+        assert!(occurrences.is_concurrent());
+
+        crossbeam::thread::scope(|scope| {
+            // Channel for sending an error from the worker threads to the main thread
+            // in the event a worker has failed.
+            let (tx_error, rx_error) = channel::bounded::<io::Error>(1);
+
+            // Channel for sending input data from the main thread to the worker threads.
+            let (tx_input, rx_input) = channel::bounded::<Option<Vec<u8>>>(2 * nthreads);
+
+            // Channel for worker threads that have errored to signal to all the other
+            // worker threads that they should stop work immediately and return.
+            let (tx_stop, rx_stop) = channel::bounded::<()>(nthreads - 1);
+
+            let mut handles = Vec::with_capacity(nthreads);
+            for _ in 0..nthreads {
+                let tx_error = tx_error.clone();
+                let rx_input = rx_input.clone();
+                let (tx_stop, rx_stop) = (tx_stop.clone(), rx_stop.clone());
+
+                let mut folder = self.clone();
+                let mut occurrences = occurrences.clone();
+
+                // Launch the worker thread...
+                let handle = scope.spawn(move |_| loop {
+                    channel::select! {
+                        recv(rx_input) -> input => {
+                            // Received from the main thread either:
+                            // * `Some(<input_data>)`, or
+                            // * `None` = a signal that no more input data will be sent.
+                            let data = match input.unwrap() {
+                                Some(data) => data,
+                                // If there is no more input data, return.
+                                None => return,
+                            };
+                            // If there is input data, process it.
+                            if let Err(e) = folder.collapse_single_threaded(&data[..], &mut occurrences) {
+                                // In the event of an error...
+
+                                // First, signal to all the other worker threads that they
+                                // should stop work immediately.  If the channel is full, it
+                                // means another thread has also errored and already sent a stop
+                                // signal to the other threads; so there is no need to wait or
+                                // to check for a `SendError` here.
+                                for _ in 0..(nthreads - 1) {
```

Comment:
I think I'd add to this comment:

> We notify all the threads about this here, rather than wait for the main input loop to see the error, so that we can also stop the input loop from iterating through the rest of the file.